src/sentry/eventstream/kafka/consumer.py

```diff
@@ -0,0 +1,295 @@
+
+    def __check_commit_log_consumer_running(self):
+        if not self.__commit_log_consumer.running():
+            try:
+                result = self.__commit_log_consumer.result(timeout=0)  # noqa
+            except TimeoutError:
+                pass  # not helpful
+
+            raise Exception('Commit log consumer unexpectedly exit!')
+
+    def __on_partition_state_change(
+            self, topic, partition, previous_state_and_offsets, current_state_and_offsets):
+        """
+        Callback that is invoked when a partition state changes.
+        """
+        logger.debug('State change for %r: %r to %r', (topic, partition),
+                     previous_state_and_offsets, current_state_and_offsets)
+
+        current_state, current_offsets = current_state_and_offsets
+        if current_offsets.local is None:
+            # It only makes sense to manipulate the consumer if we've got an
+            # assignment. (This block should only be entered at startup if the
+            # remote offsets are retrieved from the commit log before the local
+            # consumer has received its assignment.)
+            return
+
+        # TODO: This wil be called from the commit log consumer thread, so need
+        # to verify that calling the ``consumer.{pause,resume}`` methods is
+        # thread safe!
+        if current_state in (SynchronizedPartitionState.UNKNOWN, SynchronizedPartitionState.SYNCHRONIZED,
+                             SynchronizedPartitionState.REMOTE_BEHIND):
+            self.__consumer.pause([TopicPartition(topic, partition, current_offsets.local)])
+        elif current_state is SynchronizedPartitionState.LOCAL_BEHIND:
+            self.__consumer.resume([TopicPartition(topic, partition, current_offsets.local)])
+        else:
+            raise NotImplementedError('Unexpected partition state: %s' % (current_state,))
+
+    def subscribe(self, topics, on_assign=None, on_revoke=None):
+        """
+        Subscribe to a topic.
+        """
+        self.__check_commit_log_consumer_running()
+
+        def assignment_callback(consumer, assignment):
+            # Since ``auto.offset.reset`` is set to ``error`` to force human
+            # interaction on an offset reset, we have to explicitly specify the
+            # starting offset if no offset has been committed for this topic during
+            # the ``__consumer_offsets`` topic retention period.
+            assignment = {
+                (i.topic, i.partition): self.__positions.get((i.topic, i.partition)) for i in assignment
+            }
+
+            for i in self.__consumer.committed([TopicPartition(topic, partition) for (
+                    topic, partition), offset in assignment.items() if offset is None]):
+                k = (i.topic, i.partition)
+                if i.offset > -1:
+                    assignment[k] = i.offset
+                else:
+                    assignment[k] = self.initial_offset_reset(consumer, i.topic, i.partition)
```

Comment:
Ah, so `-1` is magic but you can't use it yourself? Have to request the real value?

Reply:
It gets ignored and does whatever is in `auto.offset.reset` (`error`in this case.)

It needs to be explicitly like this if the consumer group has never been seen before or has had it's registration expire since the last time it was active (more or less indistinguishable from the first scenario.)

If for some reason the consumer then tries to read parts of the log that don't exist (reads ahead or falls behind past retention) then it will error per the `auto.offset.reset` configuration because a human probably will need to make decision how to proceed.